### PR TITLE
Document Shredder Mitigation process

### PR DESCRIPTION
**Pull request by lucia-vargas-a**
Implements [DENG-4819](https://mozilla-hub.atlassian.net/browse/DENG-4819) with the documentation of shredder mitigation process for backfills.

### CODE REVIEW

**Review by kwindau — APPROVED** [review]

**Review by BenWu — APPROVED** [review]

File: src/cookbooks/data_modeling/shredder_mitigation.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+**Shredder mitigation** is a process that breaks through this risk by securely executing backfills on aggregate tables. It recalculates metrics using both existing and new data, mitigating the impact of shredder and propagating business definition changes while maintaining data integrity over time.
+
+This documentation provides step-by-step guidance with example and common scenarios, to help you achieve consistent results.
+
+For context and details on previous analysis and discussions that lead to this solution, refer to the [proposal](https://docs.google.com/document/d/1lU7qSjAOFrdcQuDbdhFzJLgP6Mkn6L8XBjRtFdq6qkU/edit?usp=sharing).
+
+[//]: # "<!-- TOC -->"
```
Comment (badboy) [inline src/cookbooks/data_modeling/shredder_mitigation.md:11]: Not sure about the intention of this piece of syntax here.
If you wanted this to be auto-filled with a table of contents, you need just this:

```suggestion
<!-- toc -->
```

File: src/cookbooks/data_modeling/shredder_mitigation.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+- Rename dimensions with known upstream changes that haven't propagated to our aggregate. To achieve this, `first_seen_year` is renamed to `first_seen_year_new` and `segment` is renamed to `segment_dau`.
+- Raise a PR to apply changes.
+- Follow the [managed backfill](https://mozilla.github.io/bigquery-etl/cookbooks/creating_a_derived_dataset/#backfilling-a-table) process with `--shredder_mitigation` parameter.
+
+## Validations
+
+Recommended data validations include:
+
+- Use SELECT EXCEPT DISTINCT to identify rows in the previous version of the table that are missing in the new version, which was just backfilled. This command performs a 1:1 comparison by checking both dimensions and metrics.
```
Comment (badboy) [inline src/cookbooks/data_modeling/shredder_mitigation.md:163]:
```suggestion
- Use `SELECT EXCEPT DISTINCT` to identify rows in the previous version of the table that are missing in the new version, which was just backfilled. This command performs a 1:1 comparison by checking both dimensions and metrics.
```

Might wanna keep keywords in code blocks.

File: src/cookbooks/data_modeling/shredder_mitigation.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+- Calculate subtotals per column, ensuring you use COALESCE for an accurate comparison of NULL values, and verify that all values match the upstream sources, except for NULL which is expected to increase.
```
Comment (badboy) [inline src/cookbooks/data_modeling/shredder_mitigation.md:164]:
```suggestion
- Calculate subtotals per column, ensuring you use `COALESCE` for an accurate comparison of `NULL` values, and verify that all values match the upstream sources, except for `NULL` which is expected to increase.
```

File: src/cookbooks/data_modeling/shredder_mitigation.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+- Prevents an accidental backfill with shredder mitigation on tables that are not set up for the process.
+- Clearly identifies which aggregate tables are set to use shredder mitigation.
+- Documents each process run, along with its purpose.
+- Provides a comprehensive set of informative and debugging messages, especially useful during first-time runs where many columns may need updating.
+
+## What to expect after running a backfill with shredder mitigation
+
+- A new version of the aggregate table is available, incorporating all new added or modified columns.
+- Totals for each column remain stable.
+- Subtotals adjusted for modified or newly added columns, with NULL values increasing by the amount corresponding to the shredded client_ids.
```
Comment (badboy) [inline src/cookbooks/data_modeling/shredder_mitigation.md:42]:
```suggestion
- Subtotals adjusted for modified or newly added columns, with `NULL` values increasing by the amount corresponding to the shredded client IDs.
```

File: src/cookbooks/data_modeling/shredder_mitigation.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+2. Metrics totals will only match if all columns with upstream changes are processed in addition to your own modifications, which can be achieved by renaming any column with upstream or local changes yet to be propagated, both in query and schema.
+
+3. Metrics totals for `NULL` data will only match if the same logic for `NULL` is applied in all versions of the query. Particularly, `country` and `city` are columns where `NULL` values are usually set to `'??'`.
+
+4. It is recommended to always run the process on a `small period of time` or a subset of data to identify potential mismatches and required changes before executing a complete backfill.
+
+#### Columns with recent changes that may require propagation to aggregates
+
+- `first_seen_date` and `first_seen_year` where the business logic for Firefox Desktop changed in 2023-Q4 to integrate data from 3 different pings.
+- `segment` where the business logic for all browsers changed in 2024-H1 to integrate the dau definition the segmentation of clients.
```
Comment (badboy) [inline src/cookbooks/data_modeling/shredder_mitigation.md:62]:
```suggestion
- `segment` where the business logic for all browsers changed in 2024-H1 to integrate the DAU definition the segmentation of clients.
```

File: src/cookbooks/data_modeling/shredder_mitigation.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+After running the backfill, you should expect the distribution of `os_version` to change, with an increase in NULL values. Additionally, column `attributed` will no longer be present, and the totals for each column will remain unchanged.
+
+### Example 1. Regular runs
+
+This example is for a table that already has the label `shredder_mitigation` and upstream changes propagated, which makes the process as simple as:
+
+- Bump the version of the table and make changes to the new version files.
+- Update the definition of `os_version` in the query and rename it to `os_version_build` both in query and schema to ensure that it is recalculated. The schema description remains valid and remove column `attributed` from the query and schema.
+- Follow the [managed backfill](https://mozilla.github.io/bigquery-etl/cookbooks/creating_a_derived_dataset/#backfilling-a-table) process using the --shredder_mitigation parameter.
```
Comment (badboy) [inline src/cookbooks/data_modeling/shredder_mitigation.md:123]:
```suggestion
- Follow the [managed backfill](https://mozilla.github.io/bigquery-etl/cookbooks/creating_a_derived_dataset/#backfilling-a-table) process using the `--shredder_mitigation` parameter.
```

File: .spelling [hunk 7]
```diff
@@ -489,3 +489,10 @@ customizable
 lookback
 rescan
 unmute
+client_ids
+os_version_build
+2023-Q4
+3
+2024-H1
+dau
+shredder_mitigation
```
Comment (badboy) [inline .spelling:498]: With the below changes you shouldn't need to list most of these as correct spellings.

```suggestion
2023-Q4
2024-H1
```

**Review by badboy — CHANGES_REQUESTED** [review]

File: src/cookbooks/data_modeling/shredder_mitigation.md (hunk 1 above)
Comment (lucia-vargas-a) [reply src/cookbooks/data_modeling/shredder_mitigation.md]: Yes, it was pending to generate the TOC, thanks!

File: .spelling (hunk 7 above)
Comment (lucia-vargas-a) [reply .spelling]: Excellent, thank you

**Review by badboy — APPROVED** [review]

**Pull request merged**